File: src/differential_equations_second_order_constant_medium.py
"""
differential_equations - second_order_constant (medium)
Generated: 2026-02-22T05:24:57.613871
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)

    if problem_type == 1:
        # Homogeneous — distinct real roots (wider range)
        r1 = nonzero(-6, 6)
        r2 = nonzero(-6, 6)
        while r2 == r1:
            r2 = nonzero(-6, 6)

        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2

        C1, C2 = symbols('C_1 C_2')
        ans = C1 * exp(r1 * t) + C2 * exp(r2 * t)

        diff_eq = f"y''{b_coeff:+d}y'{c_coeff:+d}y = 0"

        return problem(
            question=f"Find the general solution to the differential equation: ${diff_eq}$",
            answer=ans,
            difficulty=(1300, 1430),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Characteristic equation: $r^2{b_coeff:+d}r{c_coeff:+d}=0$",
                f"Factor: $(r-{r1})(r-{r2})=0$",
                f"Roots: $r_1={r1}$, $r_2={r2}$",
                f"General solution: $y = C_1 e^{{{r1}t}} + C_2 e^{{{r2}t}}$"
            ),
            answer_type="expression"
        )

    elif problem_type == 2:
        # Homogeneous — repeated root
        r = nonzero(-5, 5)

        b_coeff = -2 * r
        c_coeff = r ** 2

        C1, C2 = symbols('C_1 C_2')
        ans = (C1 + C2 * t) * exp(r * t)

        diff_eq = f"y''{b_coeff:+d}y'{c_coeff:+d}y = 0"

        return problem(
            question=f"Find the general solution to the differential equation: ${diff_eq}$",
            answer=ans,
            difficulty=(1400, 1530),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Characteristic equation: $r^2{b_coeff:+d}r{c_coeff:+d}=0$",
                f"Factor: $(r-{r})^2=0$; repeated root $r={r}$",
                f"General solution: $y = (C_1 + C_2 t)e^{{{r}t}}$"
            ),
            answer_type="expression"
        )

    elif problem_type == 3:
        # Homogeneous — complex roots
        alpha = randint(-4, 4)
        beta = randint(1, 6)

        b_coeff = -2 * alpha
        c_coeff = alpha ** 2 + beta ** 2

        C1, C2 = symbols('C_1 C_2')
        ans = exp(alpha * t) * (C1 * cos(beta * t) + C2 * sin(beta * t))

        diff_eq = f"y''{b_coeff:+d}y'{c_coeff:+d}y = 0"

        return problem(
            question=f"Find the general solution to the differential equation: ${diff_eq}$",
            answer=ans,
            difficulty=(1450, 1580),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Characteristic equation: $r^2{b_coeff:+d}r{c_coeff:+d}=0$",
                f"Using quadratic formula: $r = {alpha} \\pm {beta}i$",
                f"General solution: $y = e^{{{alpha}t}}(C_1\\cos({beta}t) + C_2\\sin({beta}t))$"
            ),
            answer_type="expression"
        )

    else:
        # Non-homogeneous: y'' + b*y' + c*y = rhs (constant)
        r1 = nonzero(-4, 4)
        r2 = nonzero(-4, 4)
        while r2 == r1:
            r2 = nonzero(-4, 4)

        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2

        rhs = nonzero(-10, 10)
        y_p = Rational(rhs, c_coeff)

        C1, C2 = symbols('C_1 C_2')
        yh = C1 * exp(r1 * t) + C2 * exp(r2 * t)
        ans = yh + y_p

        diff_eq = f"y''{b_coeff:+d}y'{c_coeff:+d}y = {rhs}"

        return problem(
            question=f"Find the general solution to the differential equation: ${diff_eq}$",
            answer=ans,
            difficulty=(1500, 1600),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Homogeneous: $y_h = C_1 e^{{{r1}t}} + C_2 e^{{{r2}t}}$",
                f"Particular (try $y_p = A$): ${c_coeff}A = {rhs}$, so $A = {latex(y_p)}$",
                f"General solution: $y = C_1 e^{{{r1}t}} + C_2 e^{{{r2}t}} + {latex(y_p)}$"
            ),
            answer_type="expression"
        )


emit(generate())

File: src/differential_equations_second_order_constant_medium.jl
# differential_equations - second_order_constant (medium)
# Generated: 2026-03-08T20:53:08.173984

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("differential_equations/second_order_constant")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Homogeneous with distinct real roots
        r1 = randint(-6, -1)
        r2 = randint(1, 6)
        while r2 == -r1
            r2 = randint(1, 6)
        end
        
        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        general_sol = "y = C_1 e^{$(r1)x} + C_2 e^{$(r2)x}"
        
        solution_steps = steps(
            "Given: \$$(diff_eq)\$",
            "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
            "Factor: \$(r - $(r1))(r - $(r2)) = 0\$",
            "Roots: \$r_1 = $(r1), r_2 = $(r2)\$",
            "Since roots are distinct and real, general solution:",
            sol("Answer", general_sol)
        )
        
        problem(
            question="Find the general solution to the differential equation: \$$(diff_eq)\$",
            answer=general_sol,
            difficulty=(1200, 1400),
            solution=solution_steps,
            time=120
        )
        
    elseif problem_type == 2
        # Homogeneous with repeated real roots
        r = randint(-5, 5)
        while r == 0
            r = randint(-5, 5)
        end
        
        b_coeff = -2*r
        c_coeff = r^2
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        general_sol = "y = C_1 e^{$(r)x} + C_2 x e^{$(r)x}"
        
        solution_steps = steps(
            "Given: \$$(diff_eq)\$",
            "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
            "Factor: \$(r - $(r))^2 = 0\$",
            "Repeated root: \$r = $(r)\$",
            "For repeated roots, general solution:",
            sol("Answer", general_sol)
        )
        
        problem(
            question="Find the general solution to the differential equation: \$$(diff_eq)\$",
            answer=general_sol,
            difficulty=(1300, 1500),
            solution=solution_steps,
            time=150
        )
        
    elseif problem_type == 3
        # Homogeneous with complex roots
        alpha = randint(-4, 4)
        beta = randint(2, 6)
        
        b_coeff = -2*alpha
        c_coeff = alpha^2 + beta^2
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        if alpha == 0
            general_sol = "y = C_1 \\cos($(beta)x) + C_2 \\sin($(beta)x)"
        else
            general_sol = "y = e^{$(alpha)x}(C_1 \\cos($(beta)x) + C_2 \\sin($(beta)x))"
        end
        
        solution_steps = steps(
            "Given: \$$(diff_eq)\$",
            "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
            "Using quadratic formula: \$r = \\frac{$(−b_coeff) \\pm \\sqrt{$(b_coeff^2 - 4*c_coeff)}}{2}\$",
            "Complex roots: \$r = $(alpha) \\pm $(beta)i\$",
            "For complex roots \$\\alpha \\pm \\beta i\$:",
            sol("Answer", general_sol)
        )
        
        problem(
            question="Find the general solution to the differential equation: \$$(diff_eq)\$",
            answer=general_sol,
            difficulty=(1400, 1600),
            solution=solution_steps,
            time=180
        )
        
    else
        # Initial value problem with distinct real roots
        r1 = randint(-5, -1)
        r2 = randint(1, 5)
        while r2 == -r1
            r2 = randint(1, 5)
        end
        
        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2
        
        y0 = randint(-10, 10)
        yp0 = randint(-10, 10)
        
        # Solve for C1 and C2
        # y(0) = C1 + C2 = y0
        # y'(0) = r1*C1 + r2*C2 = yp0
        # C1 = (yp0 - r2*y0)/(r1 - r2)
        # C2 = y0 - C1
        
        c1_num = yp0 - r2*y0
        c1_den = r1 - r2
        c1 = c1_num // c1_den
        c2 = y0 - c1
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        particular_sol = "y = $(c1)e^{$(r1)x} $(c2 >= 0 ? "+" : "")$(c2)e^{$(r2)x}"
        
        solution_steps = steps(
            "Given: \$$(diff_eq)\$, \$y(0) = $(y0)\$, \$y'(0) = $(yp0)\$",
            "Characteristic equation gives roots: \$r_1 = $(r1), r_2 = $(r2)\$",
            "General solution: \$y = C_1 e^{$(r1)x} + C_2 e^{$(r2)x}\$",
            "Apply initial conditions:",
            "\$y(0) = C_1 + C_2 = $(y0)\$",
            "\$y'(0) = $(r1)C_1 + $(r2)C_2 = $(yp0)\$",
            "Solving system: \$C_1 = $(c1), C_2 = $(c2)\$",
            sol("Answer", particular_sol)
        )
        
        problem(
            question="Solve the initial value problem: \$$(diff_eq)\$, \$y(0) = $(y0)\$, \$y'(0) = $(yp0)\$",
            answer=particular_sol,
            difficulty=(1500, 1800),
            solution=solution_steps,
            time=240
        )
    end
end